rsconnect/bundle.py

```diff
@@ -57,32 +57,39 @@
 
 
 class Manifest:
-    def __init__(self, *args, **kwargs) -> None:
-        self.data = dict()
+    def __init__(
+        self,
+        *args,
+        version: int = None,
+        environment: Environment = None,
+        app_mode: AppMode = None,
+        entrypoint: str = None,
+        quarto_inspection: dict = None,
+        image: str = None,
+        primary_html: str = None,
+        metadata: dict = None,
+        files: dict = None,
+        **kwargs
+    ) -> None:
+        self.data: dict = dict()
         self.buffer: dict = dict()
-        self._deploy_dir = None
-        version = kwargs.get("version")
-        environment = kwargs.get("environment")
-        app_mode = kwargs.get("app_mode")
-        entrypoint = kwargs.get("entrypoint")
-        quarto_inspection = kwargs.get("quarto_inspection")
-        environment = kwargs.get("environment")
-        image = kwargs.get("image")
-        primary_html = kwargs.get("primary_html")
+        self._deploy_dir: str = None
 
         self.data["version"] = version if version else 1
         if environment:
             self.data["locale"] = environment.locale
 
-        self.data["metadata"] = (
-            {
-                "appmode": app_mode.name(),
-            }
-            if app_mode
-            else {
-                "appmode": AppModes.UNKNOWN,
-            }
-        )
+        if metadata and isinstance(metadata, dict):
+            self.data["metadata"] = metadata
+        elif metadata and not isinstance(metadata, dict):
+            raise TypeError("metadata must be a dictionary.")
+        else:
+            self.data["metadata"] = {}
+            if isinstance(app_mode, AppMode):
```

Comment:
this is covered by the else in the following lines: https://github.com/rstudio/rsconnect-python/blob/c5740918da872c38cf05282de89541567347285c/rsconnect/bundle.py#L90-L91

or do we have something else in mind?

Reply:
When using arguments that default to `None`, usually the check is to see if the argument is `None` or whether a value was specified. Additional type checks usually aren't needed since the arguments are type-annotated.

Reply:
I see, I removed all the redundant isinstance checks against type annotated parameters with [99f9ac4](https://github.com/rstudio/rsconnect-python/pull/402/commits/99f9ac4b8a42fc9536fe545e15167688e1ee628b) and [0386275](https://github.com/rstudio/rsconnect-python/pull/402/commits/0386275fd89229a5b89c919d5644ab45bf04da86)